text/0080-datetime-extension.md

```diff
@@ -0,0 +1,490 @@
+It is also difficult for the calling application to predict the necessary pre-computed values that a policy writer requires for their intended purpose. These may change over time, and may also differ depending on the principal, action, and/or resource.
+
+### Additional examples
+
+**Only allow user "alice" to view JPEG photos for one week after creation.**
+
+```cedar
+permit(
+  principal == User::"alice",
+  action == PhotoOp::"view",
+  resource is Photo
+) when {
+  resource.fileType == "JPEG" &&
+  context.currentTime.since(resource.creationDate).days() <= 7
+};
+```
+
+**Allow access from a certain IP address only between 9am and 6pm UTC.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  9 <= context.currentTime.hours() &&
+  context.currentTime.hours() < 18
+};
+```
+
+**Allow access from a certain IP address only between 9am and 6pm localized to their time.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  9 <= context.currentTime.offset(principal.tzOffset).hours() &&
+  context.currentTime.offset(principal.tzOffset).hours() < 18
+};
+```
+
+**Prevent employees from accessing work documents on the weekend.**
+
+```cedar
+forbid(
+  principal,
+  action == Action::"access",
+  resource is Document
+) when {
+  [1, 7].contains(context.currentTime.dayOfWeek())
+};
+```
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"redeem",
+  resource is Prize
+) when {
+  principal.birthDate.day() == 29 &&
+  principal.birthDate.month() == 2 &&
+  context.currentTime.day() == 29 && 
+  context.currentTime.month() == 2
+};
+```
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.currentTime > datetime("20200131T23:00:00Z") &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+
+## Detailed design
+
+This RFC proposes supporting two new extension types: `datetime`, which represents a particular instant of time, up to second accuracy, and `duration` which represents a duration of time.
+To construct and manipulate these types we will provide the functions listed below.
+All of this functionality will be hidden behind a `datetime` feature flag (analogously to the current decimal and IP extensions), allowing users to opt-out if they do not want this functionality.
+
+- `datetime(s)` constructs a datetime object. Like with our other constructors, strict validation requires `s` to be a string literal, although evaluation/authorization support any string-typed expression. The string must have the form `"YYYY-MM-DDThh:mm:ssZ"` (UTC), or `"YYYY-MM-DDThh:mm:ss(+/-)hhmm"` (with Time Zone Offset).
```

Comment:
Thinking about this more, I'd vote for millisecond resolution. I think throttling is a fairly plausible use case.

Reply:
I'm fine with millisecond precision!